### PR TITLE
Shop App: View details about a product

### PR DESCRIPTION
**Pull request by VivianaSara**
## New features

  - Page View Controller

## Changes in code

- create a pageViewController for product images
- add " add to cart ", " like " and " back " buttons
- create user database

## Screenshot

- **Details about an product page**
Screenshot 2022-06-14 at 11 11 42

-  **Add to cart functionality**

> if the size is not chosen

Screenshot 2022-06-14 at 11 11 54

> if the product is already added to cart

Screenshot 2022-06-14 at 11 12 00

-  **Like functionality**

> if tap the button, the product will be added to favourites

Screenshot 2022-06-14 at 11 12 10

### CODE REVIEW

File: ShopApp/ShopApp/Controller/ImagesProductViewController.swift [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        contentImageView.image = image
+    }
+
+    func configure(index: Int, image: UIImage) {
+        self.index = index
+        self.image = image
+    }
+
+    func getIndex() -> Int { return self.index}
+
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ImagesProductViewController.swift]: empty space

File: ShopApp/ShopApp/Controller/HomeViewController.swift [hunk 2]
```diff
@@ -30,18 +32,52 @@ class HomeViewController: UIViewController {
...
+        } catch {
+            print(error)
+        }
+    }
+
     override func viewDidLoad() {
-        super.viewDidLoad()
+        async {
+            super.viewDidLoad()
+            await self.addUser()
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/HomeViewController.swift]: You can only add this line in the "async" block

File: ShopApp/ShopApp/Controller/HomeViewController.swift [hunk 3]
```diff
@@ -30,18 +32,52 @@ class HomeViewController: UIViewController {
...
+            let cellIdentifier = "CategoryCollectionViewCell"
+            let layout = UICollectionViewFlowLayout()
+            layout.itemSize = CGSize(width: UIScreen.main.bounds.width / 3, height: 120)
+            collectionView.collectionViewLayout = layout
 
-        let cellIdentifier = "CategoryCollectionViewCell"
-        let layout = UICollectionViewFlowLayout()
-        layout.itemSize = CGSize(width: UIScreen.main.bounds.width / 3, height: 120)
-        collectionView.collectionViewLayout = layout
+            collectionView.register(CategoryCollectionViewCell.nib(), forCellWithReuseIdentifier: cellIdentifier)
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/HomeViewController.swift]: Exract the rest of the code in a separate method

File: ShopApp/ShopApp/Controller/ImagesProductViewController.swift [hunk 4]
```diff
@@ -0,0 +1,29 @@
+//
+//  ImagesProductViewController.swift
+//  ShopApp
+//
+//  Created by Viviana Mesaros on 06.06.2022.
+//
+
+import UIKit
+
+class ImagesProductViewController: UIViewController {
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ImagesProductViewController.swift]: Rename it to "ProductImagesContainerViewController"

File: ShopApp/ShopApp/Controller/ProductDetailView.swift [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+//  ShopApp
+//
+//  Created by Viviana Mesaros on 06.06.2022.
+//
+
+import UIKit
+import FirebaseDatabase
+import FirebaseAuth
+
+class ProductDetailView: UIViewController {
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ProductDetailView.swift]: "ProductDetailsViewController"

File: ShopApp/ShopApp/Controller/ProductDetailView.swift [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    private var sizeSelected = ""
+    private var favorites = [String]()
+    private var cart = [String: String]()
+
+    func setProduct(product: Product) { self.product = product }
+    func setFavorites() async { self.favorites = await getFavourites() }
+    func setCart() async { self.cart = await getCart() }
+
+    // MARK: - ViewDidLoad
+    override func viewDidLoad() { async
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ProductDetailView.swift]: allign this "async" the way you did above

File: ShopApp/ShopApp/Controller/ProductDetailView.swift [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        {
+            super.viewDidLoad()
+
+            if await self.getIndex() == -1 {
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ProductDetailView.swift]: "getFavoriteProductIndex"

File: ShopApp/ShopApp/Controller/ProductDetailView.swift [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        }
+    }
+
+    // MARK: - CloseAction for close button
+    @IBAction func closeAction(sender: UIButton) {
+        dismiss(animated: true, completion: nil)
+    }
+
+    // MARK: - LikeAction for like button
+    @IBAction func likeAction(sender: UIButton) { async
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ProductDetailView.swift]: async allignment

File: ShopApp/ShopApp/Controller/ProductImagesPageViewController.swift [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        dataSource = self
+
+        // Create the first walkthrough screen
+        if let startingViewController = contentViewController(at: 0) {
+            setViewControllers([startingViewController], direction: .forward, animated: true, completion: nil)
+        }
+
+        delegate = self
+    }
+    func getCurrentIndex() -> Int { return self.currentIndex }
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ProductImagesPageViewController.swift:35]: empty space

File: ShopApp/ShopApp/Controller/ProductImagesPageViewController.swift [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    func setProduct(product: Product) { self.product = product }
+
+
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ProductImagesPageViewController.swift]: Delete them

File: ShopApp/ShopApp/Controller/HomeViewController.swift (hunk 2 above)
Comment (VivianaSara) [reply ShopApp/ShopApp/Controller/HomeViewController.swift]: I tried, but it doesn't woks the way as I want it to

File: ShopApp/ShopApp/Controller/HomeViewController.swift [hunk 11]
```diff
@@ -30,10 +32,47 @@ class HomeViewController: UIViewController {
...
+            try await refValue.updateChildValues(["name": name])
+        } catch {
+            print(error)
+        }
+    }
+
     override func viewDidLoad() {
         super.viewDidLoad()
+        async {
+            self.configureCategoryCollectionView()
```
Comment (AndreeaSburlea) [inline ShopApp/ShopApp/Controller/HomeViewController.swift]: Take this method call out of the block

File: ShopApp/ShopApp/Controller/ProductDetailsView.swift [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+//  ShopApp
+//
+//  Created by Viviana Mesaros on 06.06.2022.
+//
+
+import UIKit
+import FirebaseDatabase
+import FirebaseAuth
+
+class ProductDetailsView: UIViewController {
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ProductDetailsView.swift]: "ProductDetailsViewController"

File: ShopApp/ShopApp/Controller/ProductTableViewController.swift [hunk 13]
```diff
@@ -45,12 +43,12 @@ class ProductTableViewController: UITableViewController {
 
     // MARK: - ViewDidLoad
     override func viewDidLoad() {
+        super.viewDidLoad()
         async {
-            super.viewDidLoad()
+            await self.getAllProducts()
```
Comment (paulpop10) [inline ShopApp/ShopApp/Controller/ProductTableViewController.swift:48]: Only keep this "await" line in the "async" block

**Review by AndreeaSburlea — APPROVED** [review]

**Pull request merged**